services/user/InviteSys/src/AuthInviteSys.cpp

```diff
@@ -0,0 +1,69 @@
+
+using namespace psibase;
+using namespace UserService::Errors;
+using namespace UserService::Invite;
+using SystemService::AuthInterface;
+using SystemService::ServiceMethod;
+
+namespace UserService
+{
+   void AuthInviteSys::checkAuthSys(uint32_t                   flags,
+                                    AccountNumber              requester,
+                                    Action                     action,
+                                    std::vector<ServiceMethod> allowedActions,
+                                    std::vector<Claim>         claims)
+   {
+      auto type = flags & AuthInterface::requestMask;
+      if (type == AuthInterface::runAsRequesterReq)
+         return;  // Request is valid
+      else if (type == AuthInterface::runAsMatchedReq)
+         return;  // Request is valid
+      else if (type == AuthInterface::runAsMatchedExpandedReq)
+         abortMessage("runAs: caller attempted to expand powers");
+      else if (type == AuthInterface::runAsOtherReq)
+         abortMessage("runAs: caller is not authorized");
+      else if (type != AuthInterface::topActionReq)
+         abortMessage("unsupported auth type");
+
+      check(claims.size() == 1, onlyOneClaim.data());
+
+      auto claim = claims[0];
+      check(claim.service == SystemService::VerifyEcSys::service, mustUseVerifyEC.data());
+
+      auto inviteKey = psio::convert_from_frac<PublicKey>(claim.rawData);
+      auto inviteOpt = to<InviteSys>().getInvite(inviteKey);
+      check(inviteOpt.has_value(), inviteDNE.data());
+
+      check(action.service == InviteSys::service, restrictedService.data());
+
+      // Compile time check that acceptCreate and reject are still methods in InviteSys
+      using T1 = decltype(&UserService::Invite::InviteSys::acceptCreate);
+      using T2 = decltype(&UserService::Invite::InviteSys::reject);
+      check(action.method == "acceptCreate"_m || action.method == "reject"_m,
+            restrictedActions.data());
+
+      // Billing notes: The point of an invite is to allow those without a psibase account to create one
+      //    in a way that doesn't suffer from the chicken/egg problem that you *already* need an account
+      //    to call an action (such as accepting an invite to create an account).
+      // Therefore, this service allows the invited-sys user to be used by those without an account,
+      //    specifically for calling the accept/reject invite actions on invite-sys. That means that if
+      //    it succeeds, invited-sys is the account that will be billed for the action. Invited-sys is a
+      //    system account and has infinite resources, so the billing will always succeed.
+      // This does not mean that no one paid for the action to accept an invite, because the action is
+      //    prepaid by the user who creates the invite.
+   }
+
+   void AuthInviteSys::checkUserSys(psibase::AccountNumber user)
+   {
+      check(user == InviteSys::payerAccount, notWhitelisted.data());
```

Comment:
More to make it clear that this isn't meant for normal users, and to prevent people accidentally bricking their account, since this auth only allows you to call a couple actions.

Reply:
To make it clear that it's not intended for normal users, the easiest way is just to leave it out of the public documentation. For preventing people from accidentally bricking their accounts:
- We should use an action name that also includes an indication that this service is an auth service. checkUser is pretty generic and could mean something else in a different context.
- AuthEcSys should verify that the account has a public key set

Reply:
Okay, we can leave it out of public documentation. 
And good ideas for improving the name, and adding that verification to AuthEcSys, I'll do both of those.

Are you against adding the (admittedly not strictly necessary) restriction that this auth service can only be used by invited-sys?

Reply:
I was mainly questioning whether `checkUser` was needed at all. Given that we have it in general (and I agree that preventing users from accidentally bricking their accounts is a sufficient justification for it) we might as well keep the restriction.